datafusion/physical-expr/src/functions.rs

```diff
@@ -169,6 +169,22 @@ pub fn create_physical_expr(
                 )))))
             })
         }
+        BuiltinScalarFunction::WithTimezone => {
+            // function::return_type doesn't consider the arg value which contains the timezone info
+            // i.e. it could only output Timestamp<TimeUnit, None> or Timestamo<TimeUnit, Some(SomeFixedTimestamp))
+            // here we parse the second arg as timestamp and modify the return data_type
+            let timezone = format!("{}", input_phy_exprs[1]);
+            data_type = match data_type {
+                DataType::Timestamp(TimeUnit::Nanosecond, _) => DataType::Timestamp(TimeUnit::Nanosecond, Some(timezone)),
+                DataType::Timestamp(TimeUnit::Microsecond, _) => DataType::Timestamp(TimeUnit::Microsecond, Some(timezone)),
+                DataType::Timestamp(TimeUnit::Millisecond, _) => DataType::Timestamp(TimeUnit::Millisecond, Some(timezone)),
+                DataType::Timestamp(TimeUnit::Second, _) => DataType::Timestamp(TimeUnit::Second, Some(timezone)),
+                other => return Err(DataFusionError::Internal(format!(
+                    "Unsupported data type {other:?} as the first arg for function with_timezone",
+                )))
+            };
```

Comment:
correct the timezone in return data_type according to the args value

@alamb is this implementation ok for now? or it's recommended to modify the `return_type` function to accept some new args to achieve this?

I could submit another ticket to modify it as well